sklearn/utils/_array_api.py

```diff
@@ -0,0 +1,175 @@
+"""Tools to support array_api."""
+import numpy
+from .._config import get_config
+import scipy.special as special
+
+
+class _ArrayAPIWrapper:
+    """sklearn specific Array API compatibility wrapper
+
+    This wrapper makes it possible for scikit-learn maintainers to
+    deal with discrepancies between different implementations of the
+    Python array API standard and its evolution over time.
+
+    The Python array API standard specification:
+    https://data-apis.org/array-api/latest/
+
+    Documentation of the NumPy implementation:
+    https://numpy.org/neps/nep-0047-array-api-standard.html
+    """
+
+    def __init__(self, array_namespace):
+        self._namespace = array_namespace
+
+    def __getattr__(self, name):
+        return getattr(self._namespace, name)
+
+    def astype(self, x, dtype, *, copy=True, casting="unsafe"):
+        # Extend Array API to support `casting` for NumPy containers
+        if self._namespace.__name__ == "numpy.array_api":
+            x_np = numpy.asarray(x).astype(dtype, casting=casting, copy=copy)
+            return self._namespace.asarray(x_np)
+
```

Comment:
Here are constraints I am thinking about:

1. `numpy.array_api.Array` : I want this to have the same behavior as a `np.ndarray`. `order` is important because in principle we can pass `numpy.array_api.Array` into Cython function as long as the underlying array has the correct order.
2. `np.ndarray`: Exactly the same behavior we have now.
3. `array_api.Array`: Does not define concepts of `order` or `casting`, the best we can do is to ignore. We can not actually error, because that would restrict `_ArrayAPIWrapper` to only work with `numpy.array_api.Array`.

The question for 3 is "where do we ignore the `order` or `casting`". Right now it's in the `_ArrayAPIWrapper`, but that kind of hides it from the caller. We can bring the complexity to the caller:

```python
xp, array_api_type = get_namespace(X)
if array_api_type in {"numpy", "numpy.array_api"}:
    # Use order for numpy `ndarray` and `numpy.array_api`
    X_xp = xp.asarray(np.asarray(X, order="F"))
elif:
    # Order not supported by Array API
    X_xp = xp.asarray(X)
```

Reply:
I made a change in https://github.com/scikit-learn/scikit-learn/pull/22554/commits/0fabeab6d249e0d5eded8ab50c7aa61cb63b2e23 where I removed these helpers in the ArrayAPI wrapper. I think this makes it clearer to the caller that `ArrayAPI` does not support `order` or `casting`.